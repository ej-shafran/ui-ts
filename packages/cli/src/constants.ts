export const TEMPLATES = ["basic", "counter", "vite"];
export const DEFAULT_TEMPLATE = TEMPLATES[1];

const QUOTED_TEMPLATES = TEMPLATES.map((template) => `"${template}"`);
export const TEMPLATES_STRING =
  QUOTED_TEMPLATES.slice(0, -1).join(", ") + " or " + QUOTED_TEMPLATES.at(-1);

export const PLACEHOLDER = "PROJECT_NAME";

export const USAGE = `\
Usage:
  ui-ts [-t TEMPLATE] [PATH]

Initialize a "ui-ts" project in a given path, from one of several available templates.
If PATH is not provided, the CWD is used.

Options:
  --help, -h      Print this help and exit
  --template, -t  The template to use. Available templates are ${TEMPLATES_STRING}.
                  Default: "${DEFAULT_TEMPLATE}".`;
