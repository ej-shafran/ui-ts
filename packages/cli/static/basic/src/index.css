/* you can style your application here! */
