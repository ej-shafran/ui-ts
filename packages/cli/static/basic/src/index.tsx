import * as UI from "ui-ts";
import { app } from "./App";

UI.runApp(app);

new EventSource("/esbuild").addEventListener("change", () => location.reload());
