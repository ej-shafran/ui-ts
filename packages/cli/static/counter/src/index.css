/* you can style your application here! */

* {
  box-sizing: border-box;
  padding: 0;
  margin: 0;
}

#root,
body,
html {
  width: 100%;
  height: 100%;
}

#root {
  display: grid;
  place-items: center;
}

button {
  --bg: #2ea44f;
  --bg-dark: #1d933e;
  --border: rgba(27, 31, 35, 0.15);
  --shadow: rgba(27, 31, 35, 0.1);
  --text: white;
  --focus: rgba(46, 164, 79, 0.4);

  background-color: var(--bg);
  border: 1px solid var(--border);
  box-shadow: var(--shadow) 0 1px 0;
  color: var(--text);
  border-radius: 6px;

  cursor: pointer;
  font-size: 1rem;
  font-weight: 600;
  line-height: 1.2rem;
  padding: 0.75rem 1.5rem;
  text-align: center;
}

button:focus:not(:focus-visible):not(.focus-visible) {
  box-shadow: none;
  outline: none;
}

button:hover {
  background-color: var(--bg-dark);
}

button:focus {
  box-shadow: var(--focus) 0 0 0 3px;
  outline: none;
}
