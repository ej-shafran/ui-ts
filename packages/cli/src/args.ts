import * as E from "fp-ts/Either";
import * as TE from "fp-ts/TaskEither";
import { pipe } from "fp-ts/function";

import mri from "mri";

import { DEFAULT_TEMPLATE, TEMPLATES, USAGE } from "./constants";
import {
  UnrecognizedFlag,
  UnrecognizedTemplate,
  UserInitiated,
} from "./errors";
import { join } from "path";

export type Flags = {
  template: string;
  help: boolean | undefined;
};

const unsafeArgs = (argv: string[]) =>
  mri<Flags>(argv, {
    default: {
      template: DEFAULT_TEMPLATE,
    },
    string: ["template"],
    boolean: ["help"],
    alias: {
      t: "template",
      h: "help",
    },
    unknown(flag) {
      throw flag;
    },
  });

export const parseArgs = (argv: string[]) =>
  pipe(
    argv,
    E.tryCatchK(unsafeArgs, UnrecognizedFlag),
    TE.fromEither,
    TE.flatMap((args) =>
      !args.help
        ? TE.right(args)
        : TE.fromIOEither(() => {
            console.log(USAGE);
            return E.left(UserInitiated);
          }),
    ),
    TE.filterOrElseW(
      (args) => TEMPLATES.includes(args.template),
      (args) => UnrecognizedTemplate(args.template),
    ),
    TE.map((args) => ({
      template: args.template,
      directory: join(process.cwd(), args._[0] ?? ""),
    })),
  );
