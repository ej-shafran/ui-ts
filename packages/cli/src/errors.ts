import { ADT, match } from "ts-adt";

import { TEMPLATES_STRING, USAGE } from "./constants";

export type CLIError = ADT<{
  UserInitiated: {};
  UnrecognizedFlag: { flag: unknown };
  UnrecognizedTemplate: { template: string };
  FileExists: { path: string };
  UnknownError: { original: unknown };
}>;

export type UserInitiated = CLIError & { _type: "UserInitiated" };
export const UserInitiated: UserInitiated = {
  _type: "UserInitiated",
};

export type UnrecognizedFlag = CLIError & { _type: "UnrecognizedFlag" };
export const UnrecognizedFlag = (flag: unknown): UnrecognizedFlag => ({
  _type: "UnrecognizedFlag",
  flag: String(flag),
});

export type UnrecognizedTemplate = CLIError & { _type: "UnrecognizedTemplate" };
export const UnrecognizedTemplate = (
  template: string,
): UnrecognizedTemplate => ({
  _type: "UnrecognizedTemplate",
  template,
});

export type FileExists = CLIError & { _type: "FileExists" };
export const FileExists = (path: string): FileExists => ({
  _type: "FileExists",
  path,
});

export type UnknownError = CLIError & { _type: "UnknownError" };
export const UnknownError = (original: unknown): UnknownError => ({
  _type: "UnknownError",
  original,
});

export const handleErrors: (error: CLIError) => number = match({
  UserInitiated: () => 0,
  FileExists: ({ path }) => {
    console.log(`File already exists: "${path}"`);
    return 1;
  },
  UnrecognizedFlag: ({ flag }) => {
    console.log(USAGE);
    console.log(`\nUnrecognized flag: ${flag}`);
    return 1;
  },
  UnknownError: ({ original }) => {
    console.log("Unknown error occured. The original error was:\n", original);
    return 1;
  },
  UnrecognizedTemplate: ({ template }) => {
    console.log(
      `Unrecognized template "${template}". Available templates are ${TEMPLATES_STRING}.`,
    );
    return 1;
  },
});
