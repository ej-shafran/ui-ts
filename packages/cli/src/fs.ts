import * as TE from "fp-ts/TaskEither";
import * as A from "fp-ts/Array";
import { constVoid, flow, pipe } from "fp-ts/function";

import { cp, mkdir, readFile, readdir, writeFile } from "fs/promises";
import { join } from "path";

import { FileExists, UnknownError } from "./errors";
import { PLACEHOLDER } from "./constants";

export const isDirectoryEmpty = (dirPath: string) =>
  pipe(
    TE.tryCatch(() => mkdir(dirPath, { recursive: true }), UnknownError),
    TE.flatMapTask(() => () => readdir(dirPath)),
    TE.map(A.isEmpty),
  );

export const copyTemplate = (
  template: string,
  target: string,
  force: boolean,
) =>
  TE.tryCatch(
    () =>
      cp(join(__dirname, "..", "static", template), target, {
        force,
        errorOnExist: true,
        recursive: true,
      }),
    (error) => {
      if (
        error &&
        typeof error === "object" &&
        "code" in error &&
        error.code === "ERR_FS_CP_EEXIST"
      ) {
        return FileExists((error as { code: string; path: string }).path);
      } else {
        return UnknownError(error);
      }
    },
  );

const walkDir = (
  dirName: string,
  cb: (fileName: string) => TE.TaskEither<UnknownError, void>,
): TE.TaskEither<UnknownError, void> => {
  return pipe(
    TE.tryCatch(() => readdir(dirName, { withFileTypes: true }), UnknownError),
    TE.flatMap(
      flow(
        A.map((file) =>
          file.isDirectory()
            ? walkDir(join(dirName, file.name), cb)
            : cb(join(dirName, file.name)),
        ),
        TE.sequenceArray,
        TE.map(constVoid),
      ),
    ),
  );
};

export const replaceProjectName = (dirPath: string, projectName: string) =>
  walkDir(dirPath, (fileName) =>
    pipe(
      TE.tryCatch(() => readFile(fileName, "utf-8"), UnknownError),
      TE.map((contents) => contents.replaceAll(PLACEHOLDER, projectName)),
      TE.flatMap((contents) =>
        TE.tryCatch(() => writeFile(fileName, contents), UnknownError),
      ),
    ),
  );
