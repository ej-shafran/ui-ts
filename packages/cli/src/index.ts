#!/usr/bin/env node

import * as TE from "fp-ts/lib/TaskEither";
import { log } from "fp-ts/lib/Console";
import { pipe } from "fp-ts/function";

import { basename } from "path";

import { parseArgs } from "./args";
import { copyTemplate, isDirectoryEmpty } from "./fs";
import { promptToOverride } from "./prompt";
import { handleErrors } from "./errors";

const main = pipe(
  TE.Do,
  TE.bind("args", () => parseArgs(process.argv.slice(2))),
  TE.bindW("isEmpty", ({ args }) => isDirectoryEmpty(args.directory)),
  TE.tap(({ isEmpty, args }) =>
    !isEmpty ? promptToOverride(args.directory) : TE.right(undefined as void),
  ),
  TE.tapIO(({ args }) =>
    log(`Initializing a project in "${basename(args.directory)}"...`),
  ),
  TE.tap(({ args }) => copyTemplate(args.template, args.directory)),
  TE.match(handleErrors, () => 0),
);

main().then(process.exit);
