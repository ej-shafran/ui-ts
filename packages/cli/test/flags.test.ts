import { Left, Right, isLeft, isRight } from "fp-ts/lib/Either";
import { parseArgs } from "../src/args";
import { describe, it, expect } from "vitest";
import {
  CLIError,
  UnrecognizedFlag,
  UnrecognizedTemplate,
  UserInitiated,
} from "../src/errors";
import { DEFAULT_TEMPLATE, TEMPLATES } from "../src/constants";

const NONEXISTENT_TEMPLATE = "does not exist";
const NONEXISTENT_FLAG = "--somethingelse";

describe("positional parameter", () => {
  it("should set the `directory` option relative to the CWD", () => {
    const result = parseArgs(["../some/dir"], "cwd");

    expect(isRight(result)).toBe(true);
    expect((result as Right<{ directory: string }>).right.directory).toBe(
      "some/dir",
    );
  });
});

describe("--help, -h", () => {
  it("should print out a `help` message", () => {
    const result = parseArgs(["--help"], "");

    expect(isLeft(result)).toBe(true);
    expect((result as Left<CLIError>).left._type).toBe<CLIError["_type"]>(
      "UserInitiated",
    );
    expect((result as Left<UserInitiated>).left.log).toMatchSnapshot();
  });
});

describe("--template, -t", () => {
  it(`should default to "${DEFAULT_TEMPLATE}"`, () => {
    const result = parseArgs([], "");

    expect(isRight(result)).toBe(true);
    expect((result as Right<{ template: string }>).right.template).toBe(
      DEFAULT_TEMPLATE,
    );
  });

  it("should change the used template", () => {
    const template = TEMPLATES[1];

    const result = parseArgs(["--template", template], "");

    expect(isRight(result)).toBe(true);
    expect((result as Right<{ template: string }>).right.template).toBe(
      template,
    );
  });

  it("should fail for a nonexistent template", () => {
    const result = parseArgs(["--template", NONEXISTENT_TEMPLATE], "");

    expect(isLeft(result)).toBe(true);
    expect((result as Left<CLIError>).left._type).toBe<CLIError["_type"]>(
      "UnrecognizedTemplate",
    );
    expect((result as Left<UnrecognizedTemplate>).left.template).toBe(
      NONEXISTENT_TEMPLATE,
    );
  });
});

describe("unrecognized flag", () => {
  it("should fail", () => {
    const result = parseArgs([NONEXISTENT_FLAG], "");

    expect(isLeft(result)).toBe(true);
    expect((result as Left<CLIError>).left._type).toBe<CLIError["_type"]>(
      "UnrecognizedFlag",
    );
    expect((result as Left<UnrecognizedFlag>).left.flag).toBe(NONEXISTENT_FLAG);
  });
});
