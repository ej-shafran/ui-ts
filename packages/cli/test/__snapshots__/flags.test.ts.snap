// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`--help, -h > should print out a \`help\` message 1`] = `
"Usage:
  ui-ts [-t TEMPLATE] [PATH]

Initialize a \\"ui-ts\\" project in a given path, from one of several available templates.
If PATH is not provided, the CWD is used.

Options:
  --help, -h      Print this help and exit
  --template, -t  The template to use. Available templates are \\"basic\\", \\"counter\\" or \\"vite\\".
                  Default: \\"counter\\"."
`;
